### PR TITLE
enh(ceip): Add additional statistics including modules if present

### PR DESCRIPTION
**Pull request by lpinsivy**
<h2> Description </h2>

Add more CEIP statistics:
* Add usage of widgets in custom views
* Send OS name and version
* Check if a module can provides additional statistics

Fixes # (issue)

<h2> Type of change </h2>

Please delete options that are not relevant.

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

<h2> Target serie </h2>

Please delete series that are not relevant.

- [x] 18.10.x
- [x] 19.04.x (master)

<h2> How this pull request can be tested ? </h2>

Check data sent

<h2> Checklist </h2>

<h5> Community contributors & Centreon team </h5>

- [ ] I followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [ ] I have **rebased** my development branch on the base branch (master, maintenance).

<h5> Centreon team only </h5>

- [x] I have made sure that the **unit tests** related to the story are successful.
- [x] I have made sure that **unit tests covers 80%** of the code written for the story.
- [ ] I have made sure that **acceptance tests** related to the story are successful (**local and CI**)

### CODE REVIEW

File: cron/centreon-send-stats.php [hunk 1]
```diff
@@ -3,6 +3,7 @@
 require_once __DIR__ . '/../www/class/centreonRestHttp.class.php';
 require_once __DIR__ . '/../config/centreon-statistics.config.php';
 require_once __DIR__ . '/../www/class/centreonStatistics.class.php';
+require_once __DIR__ . '/../www/class/centreonVersion.class.php';
```
Comment (lpinsivy) [inline cron/centreon-send-stats.php]: not needed

File: www/class/centreonStatsModules.class.php [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+ * For more information : contact@centreon.com
+ *
+ */
+
+require_once _CENTREON_PATH_ . "/www/class/centreonDB.class.php";
+
+class CentreonStatsModules
+{
+    /**
+     * @var CentreonDB
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: This comment is not valid, use the code below instead
```
/**
 * @var \CentreonDB
 */
private $db;

File: www/class/centreonStatsModules.class.php [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+     * @var array InstalledModules
+     * @var array ObjectModules
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: This comment is not valid, use the code below instead.
Don't forget to name your variables in camel case.
```
/**
 * @var \ObjectModules[]
 */
private $moduleObjects;

File: www/class/centreonStatsModules.class.php [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+ *
+ */
+
+require_once _CENTREON_PATH_ . "/www/class/centreonDB.class.php";
+
+class CentreonStatsModules
+{
+    /**
+     * @var CentreonDB
+     * @var array InstalledModules
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: This comment is not valid, use the code below instead.
Don't forget to name your variables in camel case.
```
/**
 * @var \InstalledModules
 */
private $installedModules;

File: www/class/centreonStatsModules.class.php [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        if (!is_null($this->installed_modules)) {
+            return $this->installed_modules;
+        }
+        $this->installed_modules = array();
+        $stmt = $this->db->prepare("SELECT name FROM modules_informations");
+        $stmt->execute();
+        foreach ($stmt->fetchAll(PDO::FETCH_ASSOC) as $value) {
+            $this->installed_modules[] = $value['name'];
+        }
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: Why did you not return the installed modules after setting them ?

File: www/class/centreonStatsModules.class.php [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    /**
+     * Get statistics module objects
+     *
+     * @return array
+     */
+    public function getModuleObjects()
+    {
+        $this->getInstalledModules();
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: This code do not use the coding good practices. If you name a method with the 'get' prefix, it must return value 
`$installedModules = $this->getInstalledModules();`

File: www/class/centreonStatsModules.class.php [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    /**
+     * @var CentreonDB
+     * @var array InstalledModules
+     * @var array ObjectModules
+     * @var array Statistics
+     */
+    private $db;
+    private $installed_modules;
+    private $module_objects;
+    private $data;
```
Comment (callapa) [inline www/class/centreonStatsModules.class.php]: This variable is never used.

File: www/class/centreonStatsModules.class.php (hunk 5 above)
Comment (sc979) [reply www/class/centreonStatsModules.class.php]: He add an entry in `$this->installed_modules` at each occurence of the foreach
This way he returns indirectly the results

File: www/class/centreonStatsModules.class.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Get statistics from module
+     *
+     * @return array
```
Comment (sc979) [inline www/class/centreonStatsModules.class.php]:
```suggestion
     * @return array $data : The statistics of each module 
```

File: www/class/centreonVersion.class.php [hunk 9]
```diff
@@ -144,6 +144,51 @@ public function getSystem()
             $data['mysql'] = $row['Value'];
         }
 
+        return array_merge($data, $this->getVersionSystem());
+    }
+
+    /**
+     * get system information
+     *
+     * @return array
```
Comment (sc979) [inline www/class/centreonVersion.class.php]:
```suggestion
     * @return array $data An array composed with the name and version of the OS
```

File: www/class/centreonVersion.class.php [hunk 10]
```diff
@@ -144,6 +144,51 @@ public function getSystem()
...
+                $data['OS_version'] = $matches[1][1];
+            }
+        }
+        return $data;
+    }
+
+    /**
+     * Get all Centreon widgets
+     *
+     * @return array
```
Comment (sc979) [inline www/class/centreonVersion.class.php]:
```suggestion
     * @return array $data Widgets statistics
```

File: www/class/centreonStatsModules.class.php (hunk 5 above)
Comment (sc979) [reply www/class/centreonStatsModules.class.php]: Ugly, but effective

**Comment by sc979** [comment]
@callapa , you're right, I should have been more careful

File: www/class/centreonStatsModules.class.php (hunk 5 above)
Comment (lpinsivy) [reply www/class/centreonStatsModules.class.php]: This code from generation part: https://github.com/centreon/centreon/blob/master/www/class/config-generate/generate.class.php

Comment (callapa) [reply www/class/centreonStatsModules.class.php]: When you have named a method with the prefix 'get', you MUST return a value.
It's not a question of ugly or not, but of understanding the code.

**Pull request merged**